### ISSUE TITLE
MinGW-clang: Converting constructor of class with no-unique-address members overwrites last byte with zeroes 

### ISSUE THREAD

**Issue by crtrott** [llvm/llvm-project#64077]
While implementing `mdspan` I ran into an issue during MinGW CI. I reproduced this standalone on GodBolt: https://godbolt.org/z/KK8aj5bs7

The related change revision is found here: https://reviews.llvm.org/D154367

Shortish version, you have something like this:

```c++
template<class T, class M, class A>
class mdspan {
    public:
  mdspan() = default;
  mdspan(const mdspan&) = default;
  mdspan(T* ptr_, M map_, A acc_):ptr(ptr_),map(map_),acc(acc_) {}
  template<class OT, class OM, class OA>
  mdspan(const mdspan<OT, OM, OA>& other):ptr(other.ptr), map(other.map), acc(other.acc) {}

  [[no_unique_address]] T* ptr;
  [[no_unique_address]] M map;
  [[no_unique_address]] A acc;
};
```

If you now do `mdspan<const T, M2, A2>(mdspan<T, M1, A1>);` where M1, A1 and A2 are empty, but M2 is not, and A1->A2 happens via a conversion operator in A1, not a converting constructor in A2 the last byte of `ptr` gets zeroed out.

**Comment by AaronBallman** [comment]
I'm a bit confused...

`T*` certainly requires space to store it, so if `M` also requires space to store it, then the `T*` and `M` fields cannot share their storage; they cannot be potentially overlapping subobjects, right?

**Comment by crtrott** [comment]
Couple things: a) in mdspan ptr is not necessarily a T* it can very well be a class, and in fact it COULD be an empty class if you are naughty enough to exploit every possible way you can exploit customization points in mdspan. For example: accessor could be a a file system access thing, and ptr could be a handle containing a compile time string defining the file - not that I believe anyone should do that, but its possible. 

And yeah the failing case on the left hand side T* and M should NOT be overlapping since M is not empty. The one I am constructing from however had an empty M.

**Comment by crtrott** [comment]
Oh and reordering the arguments here, may very well fix my test case, but it may break another one where I test screwy things the other way around: i.e. I make M have a single size_t member every time (i.e. dextents<size_t, 1>), have an accessor with a member on the left, and an empty accessor in the original thing. It will just shift the problem to another corner case.

**Comment by tru** [comment]
Is this still something we need to fix for 17.x? I think I saw some other fixes for mdspan.

**Comment by mstorsjo** [comment]
> Is this still something we need to fix for 17.x? I think I saw some other fixes for mdspan.

I believe so. The thing that was merged was the general implementation (which was merged slightly after the 17.x branch was made, and then backported) - it has some `XFAIL` for this particular case in mingw configs, which AFAIK still are unresolved.

**Comment by crtrott** [comment]
Yeah this issue is still relevant. Note that the godbolt reproducer above does not require mdspan. It's just that it surfaces in mdspan testing, when I test legal misuse of our customization points (i.e. I test user provided implementations of customization points, which are idiosyncratic naughty, but legal).

**Comment by crtrott** [comment]
I commented on the other issue, but basically the godbolt example fails also on PowerPC with Clang 14, and presumably with clang 17 (which I am still in the process of compiling on my power platform). (Minor note: had to remove the concept and span include which were unnecessary anyway).

**Comment by mstorsjo** [comment]
This sounds a lot like #64253 - see the very minimal testcase in the comment at https://github.com/llvm/llvm-project/issues/64253#issuecomment-1665926566.

**Comment by mstorsjo** [comment]
See https://reviews.llvm.org/D157332 for a potential fix for this case.

**Comment by tru** [comment]
can this one be closed now @mstorsjo ?

**Comment by mstorsjo** [comment]
> can this one be closed now @mstorsjo ?

Yes, afaik this should have been fixed by 4d502f14aca55b8378e75373639cbee305c8c2ad.
